### PR TITLE
Refactor bokeh.core.properties to support better documentation

### LINKED ISSUE #5626: Split up properties.py
With more users writing extensions, it is now important to have better documentation about writing models, using properties, etc. The `properties.py` currently has a lot of very low level base classes, etc. that users should not really be concerned with. Propose to split off all the low level machinery into `property_bases.py` or similar, and leave `properties.py` only containing things that users might care about. Then to structure the documentation appropriately (e.g. splitting into spearete sections for basic properties and dataspecs)

### PR DESCRIPTION
**Pull request by bryevdv**
issues: closes #5626 

This is an WIP PR for some early feedback @mattpap @havocp @pzwang @philippjfr (anyone who has a little time to comment). 

This PR:

* keeps all existing tests passing
* makes everything that is actually a python descriptor be called "descriptor" in some way, and makes anything that is not a python descriptor not be called a descriptor
* now makes it easy to explain: properties are things that create descriptor for themselves (that delegate back to the property for validation, etc.).  HasProps is the thing that coordinates using properties to add those descriptors to classes.
* splits the enormous `bokeh.core.properties` module and also moves other related super low level modules into a `property` dir:
  - `bokeh.core.property.bases`
  - `bokeh.core.property.container`
  - `bokeh.core.property.descriptors`
  - `bokeh.core.property.factory`
  - `bokeh.core.property.override`
* everything left in `properties.py` is jsut the stuff users writing extensions might care typically about

Here is the relevant portion of the class hierarchy:

classes_bokeh

This seems fairly reasonable to me and makes much more sense.

More importantly, here is the module dependencies:

packages_bokeh

Note that all arrows point in the same direction (there is one circular internal dependency that does not seem to show up here) I am sure this could still be improved, suggestions very welcome. 

Things still to do:

* add a ton of docs
* get all of `bokeh.core` up to 100% coverage
* find a way to move `_sphinx_type` off the classes

### CODE REVIEW

**Comment by bryevdv** [comment]
As long as I had to be fixing up imports in charts, I went and made them all completely consistent order, style, etc. Can pretty much ignore changes in charts.

File: bokeh/models/tiles.py [hunk 1]
```diff
@@ -1,59 +1,65 @@
...
-     x_origin_offset = Float
-     y_origin_offset = Float
-     initial_resolution = Float
     """
 
     _args = ('url', 'tile_size', 'min_zoom', 'max_zoom', 'extra_url_vars')
 
     url = String("", help="""
-    tile service url (example: http://c.tile.openstreetmap.org/{Z}/{X}/{Y}.png)
+    Tile service url e.g., http://c.tile.openstreetmap.org/{Z}/{X}/{Y}.png
```
Comment (bryevdv) [inline bokeh/models/tiles.py]: This was causing a bizarre sphinx failure, cleaned up the rest of the file while I was here

File: bokeh/models/widgets/inputs.py [hunk 2]
```diff
@@ -142,11 +142,11 @@ class Slider(InputWidget):
     callback_policy = Enum(SliderCallbackPolicy, default="throttle", help="""
     When the callback is initiated. This parameter can take on only one of three options:
 
-       "continuous": the callback will be executed immediately for each movement of the slider
-       "throttle": the callback will be executed at most every ``callback_throttle`` milliseconds.
-       "mouseup": the callback will be executed only once when the slider is released.
+    * "continuous": the callback will be executed immediately for each movement of the slider
+    * "throttle": the callback will be executed at most every ``callback_throttle`` milliseconds.
+    * "mouseup": the callback will be executed only once when the slider is released.
```
Comment (bryevdv) [inline bokeh/models/widgets/inputs.py]: same, these were causing a weird sphinx error.

**Comment by bryevdv** [comment]
Test docs deployed to:

http://bokeh.pydata.org/en/test/docs/reference/core.html

This is already immensely better. Not only does this organization make things more comprehendible and approachable by itself, it also readily suggests what all the new additional docs I want to should say. 

I'd like this to go into `0.12.4` so unless there is objection I will plan to merge this Monday.

**Comment by bryevdv** [comment]
`set` and `update` on `HasProps` are identical. `update` is used while `set` does not appear to be, deprecated the latter.

**Comment by bryevdv** [comment]
removed entirely unused and untested properties `Align`, `Event`, `Function`, `This`

**Comment by bryevdv** [comment]
I also cleaned up out descriptors slightly. We were not adhering to the getter protocol by allowing `owner=None` as a default. This is remedied by simply calling `__get__` as it was intended to be called, in places where we use it explicity. I also made a detailed not about the optional `setter` argument that we added to the normal `__set__` arguments.

**Comment by bryevdv** [comment]
@mattpap I only have docstrings and ReSt file changes on this PR from here on out, and I will push them as separate commits. 

The docs are looking alot better for `bokeh.core` as are the unit tests:
```
Name                             Stmts   Miss  Cover   Missing
--------------------------------------------------------------
core/__init__                        0      0   100%
core/enums                          61      0   100%
core/has_props                     241     34    86%   112, 149, 163, 241-246, 282, 308-309, 337-338, 479, 489, 539, 560, 565, 571, 575, 578, 580, 584-585, 594, 650-651, 705-713
core/json_encoder                   70      0   100%
core/properties                    604    142    76%   177, 185-186, 196, 277-281, 298, 326, 329-334, 362-363, 395, 399, 401, 413-433, 437, 461, 464-469, 478, 482, 516-517, 521, 525-533, 545, 605-608, 617, 623, 626, 659-665, 680, 713, 756, 801, 804, 847, 850-860, 877-885, 891, 896-897, 915-923, 928-931, 934-935, 951, 978, 1105, 1143-1147, 1202-1209, 1212, 1235-1237, 1248-1250, 1255-1257, 1268-1270, 1273-1278, 1343
core/property/__init__               0      0   100%
core/property/bases                152     25    84%   49, 53, 60, 128, 139-144, 152, 226-228, 240, 246-254, 283, 287, 291, 322
core/property/containers            97      0   100%
core/property/descriptors          158      2    99%   642, 659
core/property/factory                5      0   100%
core/property/override               9      0   100%
core/property_mixins                23      0   100%
core/query                          54      0   100%
core/state                          50      0   100%
core/templates                      16      0   100%
core/validation/__init__             3      0   100%
core/validation/check               18      0   100%
core/validation/decorators          30      0   100%
core/validation/errors               3      0   100%
core/validation/warnings             3      0   100%
--------------------------------------------------------------
TOTAL                             1597    203    87%
```

**Comment by bryevdv** [comment]
Also meant to add, I plan to merge this tomorrow morning unless there is objection, then cut an RC

**Comment by bryevdv** [comment]
@mattpap several properties in `properties.py` that have their own `__init__` don't pass `serialized` or `readonly` is this an oversight?

**Comment by bryevdv** [comment]
OK, I am done for now, with this PR. There's more docs work to do (more in core, split glyphs, document.py, etc) and I'd like to add even more tests, but this is enough for now.

Reminded, test docs deployed to:

http://bokeh.pydata.org/en/test/docs/reference/core.html

File: bokeh/core/property/factory.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+.. note::
+    These classes form part of the very low-level machinery that implements
+    the Bokeh model and property system. It is unlikely that any of these
+    classes or their methods will be applicable to any standard usage or to
+    anyone who is not directly developing on Bokeh's own infrastructure.
+
+'''
+
+class PropertyFactory(object):
```
Comment (havocp) [inline bokeh/core/property/factory.py]: This used to have a subtype PropertyDescriptor which manufactured properties, and now it has a subtype Property which manufactures PropertyDescriptor. If doing that, surely this class should be renamed PropertyDescriptorFactory?

File: bokeh/core/property/bases.py [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+
+from ...util.string import nice_join
+from .containers import PropertyValueList, PropertyValueDict
+from .descriptors import BasicPropertyDescriptor
+from .factory import PropertyFactory
+
+class DeserializationError(Exception):
+    pass
+
+class Property(PropertyFactory):
```
Comment (havocp) [inline bokeh/core/property/bases.py]: yeah, you really have to rename PropertyFactory now, properties are their own factory? ;-)

**Comment by bryevdv** [comment]
I don't feel like I can review this too meaningfully, the main problem is that things were both moved (file reorg) and changed at the same time, and the diff isn't meaningful as a result.

Yes, I understand, this PR got unwieldy. I appreciate any comments that can be made, though. 

> The main change I noticed was that the names PropertyDescriptor and Property were flipped in meaning, using Property for the factory seems ok, but I'm not sure the manufactured thing should be called Descriptor (maybe BoundProperty or PropertyInstance or something along those lines?).

The original naming was definitely not good because things were called descriptors and they were not python descriptors, which confused the crap out of me any time I needed to look at things. Now I'm calling `PropertyDescriptor` things that are actually Python descriptors (i.e., it implements the descriptor protocol, i.e. has a `__get__`, `__set__`) I guess we could rename them to something else so that nothing is called "Descriptor" but at least to me, this naming is preferable. The fact that a thing is a python descriptor is the most important thing about it, it puts in in context that it is the wiring that delegates attribute access to Property instances (that's what the descriptor protocol is for).

File: bokeh/core/property/bases.py (hunk 4 above)
Comment (bryevdv) [inline bokeh/core/property/bases.py]: Agreed

**Comment by bryevdv** [comment]
ping @mattpap about this question again:

> several properties in properties.py that have their own __init__ don't pass serialized or readonly is this an oversight?

**Pull request merged**